### PR TITLE
Fix substring negative offset length

### LINKED ISSUE #33736: Function subString Inconsistent behavior between constant and table
**Describe what's wrong**
Function subString Inconsistent behavior between constant and table

**Does it reproduce on recent release?**
all

**How to reproduce**

```sql
create table t (s String) ENGINE=MergeTree() order by s;
insert into t (s) values ('12345');
select subString(s, -1, -3) from t;
select subString('12345', -1, -3);
```

**Expected behavior**

```
┌─substring(s, -1, -3)─┐
│                      │
└──────────────────────┘
┌─substring('12345', -1, -3)─┐
│                            │
└────────────────────────────┘

```

**Error message and/or stacktrace**

```
┌─substring(s, -1, -3)─┐
│ 5                    │
└──────────────────────┘
┌─substring('12345', -1, -3)─┐
│                            │
└────────────────────────────┘

```

### PR DESCRIPTION
**Pull request by rogeryk**
Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- fix substring function range error length when `offset` and `length` is negative constant and `s` is not constant.

This Closes #33736

### CODE REVIEW

**Comment by RogerYK** [comment]
The `arraySlice` and `subString` docs are are a little inaccurate.
> If you specify a negative value, the function returns an open slice `[offset, array_length - length)`.

```sql
 select arraySlice([1, 2, 3], 1, -1);
┌─arraySlice([1, 2, 3], 1, -1)─┐
│ [1,2]                        │
└──────────────────────────────┘
```
The result range is [1,2], not [1, 2).

> The ‘offset’ and ‘length’ arguments must be constants.   

The `offset` and `length` not need be constants.

File: tests/queries/0_stateless/01060_substring_negative_size.sql [hunk 1]
```diff
@@ -2,35 +2,115 @@ select substring('abcdefgh', 2, -2);
 select substring('abcdefgh', materialize(2), -2);
 select substring('abcdefgh', 2, materialize(-2));
 select substring('abcdefgh', materialize(2), materialize(-2));
+select substring(materialize('abcdefgh'), 2, -2);
```
Comment (kitaisreal) [inline tests/queries/0_stateless/01060_substring_negative_size.sql:5]: @RogerYK we prefer not to modify old tests, but create new ones.

File: tests/queries/0_stateless/00970_substring_arg_validation.reference [hunk 2]
```diff
@@ -1 +1 @@
-o
+
```
Comment (kitaisreal) [inline tests/queries/0_stateless/00970_substring_arg_validation.reference:1]: Why this and 00498 tests output changed ? How this change affect backward compatibility ?

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: tests/queries/0_stateless/01060_substring_negative_size.sql (hunk 1 above)
Comment (RogerYK) [reply tests/queries/0_stateless/01060_substring_negative_size.sql]: OK

File: tests/queries/0_stateless/00970_substring_arg_validation.reference (hunk 2 above)
Comment (RogerYK) [reply tests/queries/0_stateless/00970_substring_arg_validation.reference:1]: Some old `substring` and `arraySlice` test. If the `offset` and `length` is negative constant  and `value` is not constant, result is Incorrect.

Comment (RogerYK) [reply tests/queries/0_stateless/00970_substring_arg_validation.reference:1]: For this case
```sql
SELECT substring(materialize('hello'), -1, -1);
```
Old
```text
┌─substring(materialize('hello'), -1, -1)─┐
│ o                                        │
└─────────────────────────────────────────┘
```
New
```text
┌─substring(materialize('hello'), -1, -1)─┐
│                                         │
└─────────────────────────────────────────┘
```

**Comment by RogerYK** [comment]
@kitaisreal PTAL

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Review by alexey-milovidov — APPROVED** [review]
The changes LGTM.

**Pull request merged**